crates/ruff_index/src/list.rs

```diff
@@ -0,0 +1,234 @@
+use std::cmp::Ordering;
+use std::ops::Deref;
+
+use crate::vec::IndexVec;
+use crate::Idx;
+
+#[derive(Debug, Eq, PartialEq)]
+struct ListCell<I, K, V>(K, V, Option<I>);
+
+/// Stores one or more _association lists_, which are linked lists of key/value pairs. We
```

Comment:
> key/value pairs

Technically we don't need full key/value pairs for narrowing constraints (note that the value type is `()`).  But I plan on using this in a follow-on PR to also replace the `SmallVec`s of live bindings and declarations in `SymbolState`, which will have both keys and values to store.